plugins/field-bitmap/src/field-bitmap.js

```diff
@@ -0,0 +1,464 @@
+   * @param {Event=} e Optional mouse event that triggered the field to
+   *     open, or undefined if triggered programmatically.
+   * @param {boolean=} _quietInput Quiet input.
+   * @protected
+   */
+  showEditor_(e = undefined, _quietInput = undefined) {
+    const editor = this.dropdownCreate_();
+    Blockly.DropDownDiv.getContentDiv().appendChild(editor);
+    Blockly.DropDownDiv.showPositionedByField(
+        this, this.dropdownDispose_.bind(this));
+  }
+
+  /**
+   * Updates the block display and editor dropdown when the field re-renders.
+   * @protected
+   * @override
+   * @return void
+   */
+  render_() {
+    super.render_();
+
+    if (!this.getValue()) {
+      return;
+    }
+
+    if (this.blockDisplayPixels_) {
+      this.forAllCells_((r, c) => {
+        const pixel = this.getValue()[r][c];
+
+        if (this.blockDisplayPixels_) {
+          this.blockDisplayPixels_[r][c].style.fill =
+            pixel ? FILLED_PIXEL_COLOR : EMPTY_PIXEL_COLOR;
+        }
+        if (this.editorPixels_) {
+          this.editorPixels_[r][c].style.background =
+            pixel ? FILLED_PIXEL_COLOR : EMPTY_PIXEL_COLOR;
+        }
+      });
+    }
+  }
+
+  /**
+   * Determines whether the field is editable
+   * @return true since it is always editable. */
+  updateEditable() {
+    return true;
+  }
+
+  /**
+   * Creates the slider editor and add event listeners.
+   * @return {!Element} The newly created dropdown menu.
+   * @private
+   */
+  dropdownCreate_() {
+    const dropdownEditor =
+        this.createElementWithClassname_('div', 'dropdownEditor');
+    const pixelContainer =
+        this.createElementWithClassname_('div', 'pixelContainer');
+    dropdownEditor.appendChild(pixelContainer);
+
+    this.bindEvent_(dropdownEditor, 'mouseup', this.onMouseUp_);
+    this.bindEvent_(dropdownEditor, 'mouseleave', this.onMouseUp_);
+
+    this.editorPixels_ = [];
+    for (let r = 0; r < this.imgHeight_; r++) {
+      this.editorPixels_.push([]);
+      const rowDiv = this.createElementWithClassname_('div', 'pixelRow');
+      for (let c = 0; c < this.imgWidth_; c++) {
+        // Add the button to the UI and save a reference to it
+        const button = this.createElementWithClassname_('div', 'pixelButton');
+        this.editorPixels_[r].push(button);
+        rowDiv.appendChild(button);
+
+        // Handle clicking a pixel
+        this.bindEvent_(button, 'mousedown', () => {
+          this.onMouseDownInPixel_(r, c);
+          return true;
+        });
+
+        // Handle dragging into a pixel when mouse is down
+        this.bindEvent_(button, 'mouseenter', () => {
+          this.onMouseEnterPixel_(r, c);
+        });
+      }
+      pixelContainer.appendChild(rowDiv);
+    }
+
+    // Add control buttons below the pixel grid
+    this.addControlButton_(dropdownEditor, 'Randomize', this.randomizePixels_);
+    this.addControlButton_(dropdownEditor, 'Clear', this.clearPixels_);
+
+    this.render_();
```

Comment:
I'm confused by this call to render.

Reply:
Was originally needed to set the initial values of the editor pixels. Instead, now it sets the value during their creation.